python/GafferUI/RandomUI.py

```diff
@@ -175,29 +175,43 @@
 
 class _RandomColorPlugValueWidget( GafferUI.PlugValueWidget ) :
 
+	__gridSize = imath.V2i( 10, 3 )
+
 	def __init__( self, plug, **kw ) :
 
 		self.__grid = GafferUI.GridContainer( spacing = 4 )
 
 		GafferUI.PlugValueWidget.__init__( self, self.__grid, plug, **kw )
 
 		with self.__grid :
-			for x in range( 0, 10 ) :
-				for y in range( 0, 3 ) :
+			for x in range( 0, self.__gridSize.x ) :
+				for y in range( 0, self.__gridSize.y ) :
 					GafferUI.ColorSwatch( parenting = { "index" : ( x, y ) } )
 
-		self._updateFromPlug()
-
-	def _updateFromPlug( self ) :
+	@staticmethod
+	def _valuesForUpdate( plugs ) :
 
-		node = self.getPlug().source().node()
+		node = next( iter( plugs ) ).source().node()
 		seed = node["seed"].getValue()
 
-		gridSize = self.__grid.gridSize()
-		for x in range( 0, gridSize.x ) :
-			for y in range( 0, gridSize.y ) :
-				self.__grid[x,y].setColor( node.randomColor( seed ) )
+		result = []
+		for x in range( 0, _RandomColorPlugValueWidget.__gridSize.x ) :
+			column = []
+			for y in range( 0, _RandomColorPlugValueWidget.__gridSize.y ) :
+				column.append( node.randomColor( seed ) )
 				seed += 1
+			result.append( column )
+
+		return result
+
+	def _updateFromValues( self, values, exception ) :
+
+		for x in range( 0, self.__gridSize.x ) :
+			for y in range( 0, self.__gridSize.y ) :
+				if exception is not None :
+					self.__grid[x,y].setColor( imath.Color3f( 1, 0.33, 0.33 ) )
+				elif len( values ) :
+					self.__grid[x,y].setColor( values[x][y] )
```

Comment:
I don't really like this silently ignores a zero length `values`.  When do we anticipate this?  Would it not be better to error?

Reply:
This is because of https://github.com/GafferHQ/gaffer/pull/5055/files#diff-dde4f13eeee63f106166684565203352b00dfe158fffa653745d2129b3299a56R550-R553. For most widgets, it "just works" to do `updateFromValues( [] )` prior to a background update, as it puts the widget into a state that visually makes sense as "no values" or "stale values". But that's not something that falls out naturally for the RandomUI. There were one or two other widgets that had a similar issue, that will come up in a future PR.

I suppose we could introduce a new visual state to the RandomUI to show that the values are pending, but it's usually quick enough that it would just be visual noise. So I did the easy thing and just skipped the update.

One related thing I was considering was to use a specific sentinel empty list for this pre-call update. So you could test `values is PlugValueWidget.pendingValues` or similar, to know the reason for having empty values. If we had that, I could use it in an explanatory assertion, rather than skip the update entirely.

Reply:
I guess it makes sense that we're skipping the pending value indicator here because we know updates will be fast, but it would be good to have a comment that this is what's happening.

Reply:
Added comment in https://github.com/GafferHQ/gaffer/pull/5055/commits/d2819b7222cd68c8dc688469fda83f7599b753ec.